testing/src/test/java/io/stargate/it/storage/StargateContainer.java

```diff
@@ -0,0 +1,528 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.it.storage;
+
+import static io.stargate.starter.Starter.STARTED_MESSAGE;
+import static java.lang.management.ManagementFactory.getRuntimeMXBean;
+
+import io.stargate.it.storage.StargateParameters.Builder;
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.ServerSocket;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.commons.exec.CommandLine;
+import org.apache.commons.exec.DefaultExecutor;
+import org.apache.commons.exec.ExecuteWatchdog;
+import org.apache.commons.exec.Executor;
+import org.apache.commons.exec.LogOutputStream;
+import org.apache.commons.exec.PumpStreamHandler;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * JUnit 5 extension for tests that need a Stargate OSGi container running in a separate JVM.
+ *
+ * <p>Note: this extension requires {@link ExternalStorage} to be activated as well. It is
+ * recommended that test classes be annotated with {@link UseStargateContainer} to make sure both
+ * extensions are activated in the right order.
+ *
+ * <p>Note: this extension does not support concurrent test execution.
+ *
+ * @see StargateSpec
+ * @see StargateParameters
+ */
+public class StargateContainer
+    implements BeforeAllCallback,
+        AfterAllCallback,
+        BeforeEachCallback,
+        AfterEachCallback,
+        ParameterResolver {
+  private static final Logger LOG = LoggerFactory.getLogger(StargateContainer.class);
+
+  private static final File LIB_DIR = new File(System.getProperty("stargate.libdir"));
+  private static final int PROCESS_WAIT_MINUTES =
+      Integer.getInteger("stargate.test.process.wait.timeout.minutes", 10);
+
+  private static final AtomicBoolean suiteExecuting = new AtomicBoolean();
+  private static final AtomicBoolean testExecuting = new AtomicBoolean();
+
+  // the first 10 addresses are reserved for storage nodes
```

Comment:
I see one problem with limiting it to 10 nodes only for SG. 
Assuming that we will have 3 test classes that start 3 stargate nodes, and the rest of the test classes start 1 SG node. If the ordering of the tests will be unfortunate, that is:
- test with 3 nodes
- test with 1 node
- test  with 3 nodes
- test with 1 node
- test with 3 nodes
We will start and stop 12 SG nodes, exceeding this limit of 10 nodes. Right now it is not a problem, but maybe in the future?
Can we increase this limit of addresses reserved for SG?

Reply:
10 addresses are reserved for C* or DSE, which we do not run concurrently. I do not imagine us needing more than 10 nodes of C* for tests :thinking: